src/utils/generalFunctions.js

```diff
@@ -67,14 +68,93 @@ export function aboutPopup(call, classes = {}) {
   });
 }
 
+/* breaks a tool id into the alphabetical and numerical components */
+function breakToolId(tabId) {
+  const index = tabId.search(/\d/);
+  const alpha = tabId.substring(0, index - 1);
+
+  if (index < 0)
+    return [alpha];
+
+  const number = new Number(tabId.substring(index));
+
+  return [alpha, number];
+}
+
+/* get the structure from the tabsMap that will allow us
+ * to efficently organize tool ids.
+ */
+function computeIds(tabsMap) {
+  const res = {};
+
+  /* get a map of this format:
+   * {
+   *   Topics: { last: 3, busy: [2] },
+   *   VarsDebug: { last 2, busy: [1] },
+   * }
+   */
+  for (const [tabId] of tabsMap) {
+    const [alpha, number] = breakToolId(tabId);
+    const index = tabId.search(/\d/);
+    if (index < 0)
+      continue;
+    const specific = res[alpha] ?? { last: 0, busy: {}, free: [] };
+    specific.busy[number] = 1;
+    if (number > specific.last)
+      specific.last = number;
+    res[alpha] = specific;
+  }
+
+  /* turn it into:
+   * {
+   *   Topics: { last: 3, free: [1] },
+   *   VarsDebug: { last 2, free: [] },
+   * }
+   */
+  for (const toolName in res) {
+    const specific = res[toolName];
+    for (let j = 0; j < specific.last; j++)
+      if (!specific.busy[j])
+        specific.free.push(j);
+    delete specific.busy;
+    specific.last++;
+  }
+
+  return res;
+}
+
+const workspace = new Workspace();
+const tabsMap = workspace.getTabs();
+const toolIds = computeIds(tabsMap);
```

Comment:
I hate that we are just floating around in the middle of this `generalFunctions` file. I think you might be able to come up with some better way to do this.
Also, maybe all this "tool tab" logic should be in its own file, since none of these are actual "general functions" but very specific ones.

Reply:
This code is only used here. We already have too many files.